File: javascript/say/say.js
module.exports = {
  inEnglish(input) {
    if (input < 0 || input > 999999999999) {
      throw new Error('Number must be between 0 and 999,999,999,999.');
    }
    // Only one reason to say zero.
    if (input == 0) { return 'zero'; }

    let parts = [];
    let sinput = input.toString();
    let iteration = 0;
    // Return an array of three part decimals to say, plus the suffix
    // say at the end of each part.
    for (var x=sinput.length; x>0; x=x-3) {
      let start = (x-3 < 0) ? 0 : x-3;
      parts.unshift({
        suffix: suffixes[iteration++],
        number: parseInt(sinput.slice(start, x), 10)
      });
    }

    let text = [];
    parts.forEach((part) => {
      // We don't need to say anything and don't want the suffix
      // later.
      if (part.number == 0) { return; }

      // Say the hundres part if we have one.
      let hundreds = Math.floor(part.number/100);
      if (hundreds > 0) {
        text.push(this.sayOnes(hundreds));
        text.push('hundred');
        part.number -= hundreds * 100;
      }

      // Say the tens part if we have one.
      let tens = Math.floor(part.number/10);
      if (tens >= 2) {
        text.push(this.sayTens(tens * 10));
        part.number -= (tens * 10);
        if (part.number > 0) {
          text.push('-');
        }
      }
      text.push(this.sayOnes(part.number));
      text.push(part.suffix);
    });

    // Stitch it all into text, hyphens pull words together.
    return text.join(' ').trim().replace(/ - /g, '-');
  },

  sayTens(input) {
    let out = '';
    tens.forEach((pair) => {
      let [n, r] = pair;
      if (n == input) { out = r; }
    });
    return out;
  },

  sayOnes(input) {
    let out = '';
    numerals.forEach((pair) => {
      let [n, r] = pair;
      if (n == input) { out = r; }
    });
    return out;
  }
};

const numerals = [
  [ 19, 'nineteen'  ],
  [ 18, 'eighteen'  ],
  [ 17, 'seventeen' ],
  [ 16, 'sixteen'   ],
  [ 15, 'fifteen'   ],
  [ 14, 'fourteen'  ],
  [ 13, 'thirteen'  ],
  [ 12, 'twelve'    ],
  [ 11, 'eleven'    ],
  [ 10, 'ten'       ],
  [  9, 'nine'      ],
  [  8, 'eight'     ],
  [  7, 'seven'     ],
  [  6, 'six'       ],
  [  5, 'five'      ],
  [  4, 'four'      ],
  [  3, 'three'     ],
  [  2, 'two'       ],
  [  1, 'one'       ]
];

const tens = [
  [ 100, 'hundred'  ],
  [  90, 'ninety'   ],
  [  80, 'eighty'   ],
  [  70, 'seventy'  ],
  [  60, 'sixty'    ],
  [  50, 'fifty'    ],
  [  40, 'forty'    ],
  [  30, 'thirty'   ],
  [  20, 'twenty'   ]
];

const suffixes = [
  '',
  'thousand',
  'million',
  'billion',
];

File: python/say/say.py
NUMBERS = {
    1: 'one',
    2: 'two',
    3: 'three',
    4: 'four',
    5: 'five',
    6: 'six',
    7: 'seven',
    8: 'eight',
    9: 'nine',
    10: 'ten',
    11: 'eleven',
    12: 'twelve',
    13: 'thirteen',
    14: 'fourteen',
    15: 'fifteen',
    16: 'sixteen',
    17: 'seventeen',
    18: 'eighteen',
    19: 'nineteen',
    20: 'twenty',
    30: 'thirty',
    40: 'forty',
    50: 'fifty',
    60: 'sixty',
    70: 'seventy',
    80: 'eighty',
    90: 'ninty'
}

CHUNKS = [
    # { 'word': 'trillion', 'number': 1e12 },
    { 'word': 'billion', 'number': 1e9 },
    { 'word': 'million', 'number': 1e6 },
    { 'word': 'thousand', 'number': 1e3 },
    { 'word': '', 'number': 1e0 }
]

def say(number):
    if number < 0 or number >= 1e12:
        raise AttributeError

    if number == 0:
        return 'zero'

    number_string = ''
    say_and = False

    for c in CHUNKS:
        part = number // c['number']
        if part > 0:
            number_string += say_triplet(part) + ' ' + c['word'] + ' '
        else:
            if len(number_string) > 1:
                say_and = True

        number -= part * c['number']

        if number > 0 and len(number_string) > 0 and say_and:
            number_string += 'and '
            say_and = False

    return number_string.strip()


def say_triplet(number):
    number_string = ''
    seperator = ' '
    hundreds_part = number//100

    if hundreds_part > 0:
        number_string += say_triplet(hundreds_part) + ' hundred'
        number = number - hundreds_part * 100
        if number > 0:
            number_string += ' and'

    while number > 0:
        largest_part = max([i for i in NUMBERS.keys() if i <= number])

        number_string += seperator + NUMBERS[largest_part]

        if 20 <= largest_part <= 90:
            seperator = '-'
        else:
            seperator = ' '

        number -= largest_part

    return number_string.strip()